### PR TITLE
feat: #560 google map web component

### PR DESCRIPTION
**Pull request by trankhacvy**

### CODE REVIEW

File: packages/web-components/src/common/utils/map-helper.ts [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+      map,
+    })
+    const address = {
+      line1: property?.address?.line1 || '',
+      line2: property?.address?.line2 || '',
+    }
+    const lettingPrice = property?.letting?.rent
+    const rentFrequency = property?.letting?.rentFrequency
+
+    const bedrooms = property && property.bedrooms
```
Comment (tanphamhaiduong) [inline packages/web-components/src/common/utils/map-helper.ts]: can change like `property?.bedrooms`

File: packages/web-components/src/common/utils/map-helper.ts [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    const bathrooms = property && property.bathrooms
```
Comment (tanphamhaiduong) [inline packages/web-components/src/common/utils/map-helper.ts]: can change like `property?. bathrooms `

File: packages/web-components/jest.config.js [hunk 3]
```diff
@@ -13,9 +13,9 @@ module.exports = {
   coverageThreshold: {
     global: {
       branches: 60,
-      functions: 90,
-      lines: 98,
-      statements: 97,
+      functions: 70,
```
Comment (willmcvay) [inline packages/web-components/jest.config.js]: That's a big drop in test coverage - we should work on this

File: packages/web-components/src/common/utils/__mocks__/mock-google-map.ts [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+
+  const maps = {
+    Animation: {
+      BOUNCE: 1,
+      DROP: 2,
+      Lo: 3,
+      Go: 4,
+    },
+    BicyclingLayer: jest.fn().mockImplementation(function() {
+      // @ts-ignore
```
Comment (willmcvay) [inline packages/web-components/src/common/utils/__mocks__/mock-google-map.ts]: Why all the ts-ignores - surely the mock should conform to the type?

File: packages/web-components/src/common/utils/map-helper.ts [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  bedrooms?: number
+  bathrooms?: number
+  marketingMode?: string
+  lettingPrice?: number
+  rentFrequency?: string
+  imageUrl?: string
+  price: string
+}
+
+export const getContent = ({
```
Comment (willmcvay) [inline packages/web-components/src/common/utils/map-helper.ts]: Does this not make sense as a svelte component? I get why you wouldn't when it was a React component but Svelte is just HTML and CSS - we could get rid of the inline styles too

File: packages/web-components/src/scripts/rollup.config.google-map-widget.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import svelte from 'rollup-plugin-svelte'
```
Comment (willmcvay) [inline packages/web-components/src/scripts/rollup.config.google-map-widget.js]: Don't think this file is needed anymore

File: packages/web-components/src/search-widget/client/components/__tests__/google-map.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import GoogleMap from '../google-map.svelte'
+import { render } from '@testing-library/svelte'
+import { get } from 'svelte/store'
+// import { loader } from '../../../../common/utils/loader'
```
Comment (willmcvay) [inline packages/web-components/src/search-widget/client/components/__tests__/google-map.ts]: Can be removed

File: packages/web-components/src/search-widget/client/components/google-map.svelte [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+    if (map && storeInstance && storeInstance.properties !== store.properties) {
+      showPropertiesMarker(map, store, searchWidgetStore)
+      storeInstance = store
+    }
+
+    storeInstance = store
+  })
+
+  onMount(() => {
+    (window as any).onMapReady = () => {
```
Comment (willmcvay) [inline packages/web-components/src/search-widget/client/components/google-map.svelte]: I left this window hack in but should be removed - you can do
```
Object.defineProperty(window, 'onMapReady', {
  value: // callback here!
})
```

File: packages/web-components/src/search-widget/client/components/google-map.svelte [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  
+  const searchWidgetStore = SearchWidgetStore.default
+
+  let storeInstance: SearchWidgetStore.SearchWidgetStore
+  let map: google.maps.Map
+  let mapElement: HTMLDivElement
+  let mapLoading: boolean = false
+
+  const unsubscribeSearchWidgetStore = searchWidgetStore.subscribe(store => {
+    if (map && storeInstance && storeInstance.properties !== store.properties) {
```
Comment (willmcvay) [inline packages/web-components/src/search-widget/client/components/google-map.svelte]: This is a note to me - I am not sure this is the best solution but will try and come up with something better - leave for now!

File: packages/web-components/src/common/utils/__mocks__/mock-google-map.ts [hunk 10]
```diff
@@ -0,0 +1,386 @@
...
+            }
+          },
+        }
+      })
+      .mockName('addListener')
+
+    createMockFuncsFromArray(instance, ['bindTo', 'get', 'notify', 'set', 'setValues', 'unbind', 'unbindAll'])
+  }
+
+  const maps = {
```
Comment (willmcvay) [inline packages/web-components/src/common/utils/__mocks__/mock-google-map.ts]: This mock gets loaded in the jest-setup file at the root of the project - might be not necessary at all!

File: packages/web-components/src/search-widget/client/core/store.ts [hunk 11]
```diff
@@ -1,11 +1,18 @@
...
 
 export interface SearchWidgetStore {
   initializers: SearchWidgeInitializers
   properties: PagedResultPropertyModel_ | null
   propertyImages: Record<string, PropertyImageModel> | null
+  selectedProperty: PropertyModel | null
+  searchType?: SearchType
+  searchKeyword?: string
+  markers: google.maps.Marker[]
+  mapLoading: boolean
```
Comment (willmcvay) [inline packages/web-components/src/search-widget/client/core/store.ts]: I think mapLoading is not now referenced in the code, if not, we can remove

File: packages/web-components/src/search-widget/client/utils/__tests__/google-map.ts [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+      selectedProperty: null,
+      markers: [],
+      mapLoading: false,
+    }
+
+    showPropertiesMarker(maps, storeInstance, searchWidgetStore)
+    expect(createMarker).toHaveBeenCalledTimes(1)
+  })
+
+  // it('should return correctly when no properties', () => {
```
Comment (willmcvay) [inline packages/web-components/src/search-widget/client/utils/__tests__/google-map.ts]: See my comment about tests - we should test this function better - it is quite complex - recommend as a start point breaking down into blocks and testing them separately. Happy to help with this

File: packages/web-components/src/search-widget/client/utils/google-map.ts [hunk 13]
```diff
@@ -0,0 +1,55 @@
+import { createMarker, getLatLng } from '../../../common/utils/map-helper'
+import { DEFAULT_ZOOM } from '../../../common/utils/constants'
+import { SearchWidgetStore } from '../core/store'
+import { Writable } from 'svelte/store'
+
+export const showPropertiesMarker = (
```
Comment (willmcvay) [inline packages/web-components/src/search-widget/client/utils/google-map.ts]: See my previous - it's quite complex!

**Pull request merged**